runtime-testsuite/src/org/antlr/v4/testgen/TestGenerator.java

```diff
@@ -30,202 +30,315 @@
+        showOptionMulti("-exclude  ", excludes);
+        showOptionMulti("-target   ", targets);
+    }
+
+    protected void showOptionMulti(String flag, Collection<? extends Object> value) {
+        info(value.isEmpty() ? " no "+flag
+                             : "    "+flag+" "+join(", ", value));
+    }
+
+    public void execute() {
+        for (String target : targets) {
+            genTarget(target);
+		} 
+    }
 
+	public void genTarget(String target) {
+        targetName = target;
+		File targetOutputDir = outputDir(targetName);
+		info(String.format("Generating target %s => %s", targetName, targetOutputDir));
+
+        File targetTemplatesRoot = new File(rootDir, "resources/org/antlr/v4/test/runtime");
+        File targetTemplateFile = targetTemplateFile(targetTemplatesRoot, targetName);
+
+		STGroup targetGroup = new STGroupFile(targetTemplateFile.getPath());
+        targetGroup.setListener(errorBuffer);
 
-	public void execute() {
-		STGroup targetGroup = new STGroupFile(runtimeTemplate.getPath());
 		targetGroup.registerModelAdaptor(STGroup.class, new STGroupModelAdaptor());
 		targetGroup.defineDictionary("escape", new JavaEscapeStringMap());
 		targetGroup.defineDictionary("lines", new LinesStringMap());
 		targetGroup.defineDictionary("strlen", new StrlenStringMap());
-		generateCodeForFoldersInIndex(targetGroup);
+
+		generateCodeForFoldersInIndex(targetGroup, targetOutputDir);
+
+        if (errorBuffer.errors.size() > 0)
+            throwTemplateError();
 	}
 
-	protected void generateCodeForFoldersInIndex(STGroup targetGroup) {
-		File targetFolder = getOutputDir(testTemplates+"");
-		STGroup index = new STGroupFile(testTemplates+"/Index.stg");
-		index.load(); // make sure the index group is loaded since we call rawGetDictionary
-		Map<String, Object> folders = index.rawGetDictionary("TestFolders");
-		if (folders != null) {
-			for (String key : folders.keySet()) {
-				final String testDir = testTemplates + "/" + key;
-				STGroup testIndex = new STGroupFile(testDir + "/Index.stg");
-				testIndex.load();
-				Map<String, Object> templateNames = testIndex.rawGetDictionary("TestTemplates");
-				if ( templateNames != null && !templateNames.isEmpty() ) {
-					final ArrayList<String> sortedTemplateNames = new ArrayList<String>(templateNames.keySet());
-					Collections.sort(sortedTemplateNames);
-					generateTestFile(testIndex, targetGroup, testDir, sortedTemplateNames, targetFolder);
-				}
-			}
-		}
+	protected void generateCodeForFoldersInIndex(STGroup targetGroup, File targetOutputDir) {
+        Collection<String> testFolders = new ArrayList<String>(20);
+        Collection<String> excludeFolders = new TreeSet<String>();
+
+		STGroup rootIndex = new STGroupFile(new File(testTemplatesRoot, "Index.stg").getPath());
+
+        // let output file line endings be the same kind found in top Index.stg
+        lineSeparator = detectLineSeparator(rootIndex.getFileName(), lineSeparator);
```

Comment:
💭 Now that we have **.gitattributes**, this should not matter. Git will normalize line endings anyway when the source files are committed.